utils/zerotrie/src/serde.rs

```diff
@@ -132,7 +139,10 @@ where
             lm.serialize(serializer)
         } else {
             let bytes = self.as_bytes();
-            bytes.serialize(serializer)
+            let mut all_in_one_vec = Vec::with_capacity(bytes.len() + 1);
+            all_in_one_vec.push(Self::FLAGS);
+            all_in_one_vec.extend(bytes);
+            all_in_one_vec.serialize(serializer)
```

Comment:
this can be done without allocating a vec, also below

```suggestion
            let mut seq = serializer.serialize_seq(Some(bytes.len() + 1))?;
            seq.serialize_element(Self::FLAGS);
            for b in self.bytes {
                seq.serialize_element(b)?;
            }
            seq.end()
```

Reply:
Actually I should be explicitly using `serialize_bytes`. I don't want `serialize_seq`

Reply:
I think postcard represents serialize_bytes and serialize_seq the same way, but I think that's not a guarantee, and on the other end we definitely want deserialize_borrowed_bytes

Reply:
I'm not sure `serialize_bytes` is better than `serialize_seq` if you have to reallocate the whole vector in order to use it.

Reply:
The [contract of `serialize_bytes`](https://docs.rs/serde/latest/serde/trait.Serializer.html#tymethod.serialize_bytes) says:

> Enables serializers to serialize byte slices more compactly or more efficiently than other types of slices. 

On the other end I want to use [`deserialize_bytes`](https://docs.rs/serde/latest/serde/trait.Deserializer.html#tymethod.deserialize_bytes). I think this was working with Postcard because it happens to have the same representation, but I imagine other Serde impls could have different representations for "bytes" versus "seq of u8". We should probably test our code with something other than Postcard.

Reply:
Indeed, it breaks if I use `serialize_seq` with `deserialize_bytes` in the `bson` crate:

> called `Result::unwrap()` on an `Err` value: DeserializationError { message: "invalid type: sequence, expected a borrowed byte array" }

I would add a test case, except the `bson` crate pulls a bunch of new crates into the Cargo.lock file. To reproduce, add this to any of the test functions such as `test_serde_simpleascii_cow`:

```rust
let bson_bytes = bson::to_vec(&original).unwrap();
let bson_recovered: ZeroTrieSimpleAsciiCow = bson::from_slice(&bson_bytes).unwrap();
```

Reply:
I found another serde crate with the same behavior, `rmp-serde`, and added a test with it.

Reply:
Upstream Serde issue about `&[u8]` using serialize_seq instead of serialize_bytes:

https://github.com/serde-rs/serde/issues/2474